technote: tic GitHub Actions support (#621)
ropensci/roweb2: content/technotes/2020-02-26-tic-ghactions.Rmd

### Addtional small tweaks

- {tic} always runs `R CMD javareconf` to ensure a working installation of {rJava}
- {tic} uses four cores for installing dependencies instead of just running sequentially
- {tic} sets the env var `RGL_USE_NULL` to make it possible to install the {rgl} package if needed
- {tic} installs all packages from source on Linux instead of using the R package binaries provided by RStudio.
  This is to mimic the normal behavior on Linux and ensure a proper linking against system libraries.

## Getting Started

If you are triggered now to try out {tic} and GitHub Actions, here is how to get started:

```{r, eval=FALSE}
library("tic")

use_ghactions_deploy() # (optional) setup deployment from builds
use_ghactions_yml() # use deploy = TRUE to deploy from builds
use_tic_r() # add tic.R (replace `ci_on_travis()` by `ci_on_ghactions()`)
use_tic_badge("ghactions") # (optional) add a badge to README.md/README.Rmd
```

Alternatively you can also run `tic::use_tic()` to start an interactive wizard guiding you through the setup process.

After that, commit `.github/workflows/main.yml` and `tic.R` and have a look at the "Actions" pane in the repository.

## Summary

> {tic} aims to simplify CI related tasks for R projects (not just packages by the way) and now provides supercharged support for GitHub Actions.

Reviewer: I think at some point in the article I'd love to see two config files side by side or even a table of steps with one column for r-lib/actions and one for the tic version.

Author: While I see why its nice having config files directly in the post, I fear that they might get outdated and people will start with these by c/p them from the blog post. 

Maybe putting them into a "details" block with a warning to not c/p them?

Reviewer: Or maybe a summary of current key differences in the conclusion?

Author: I added a paragraph in the summary section: https://github.com/ropensci/roweb2/pull/621/commits/f47810c5aa4709592021d8811d7947f23ede05fd

> {tic} aims to simplify CI related tasks for R projects (not just packages by the way) and now provides supercharged support for GitHub Actions.

Reviewer: In the conclusion re-acknowledge that the GH Actions support is built on top of r-lib/actions?

If you encounter problems, please have a look at the vignettes of {tic}, especially the "Troubleshooting" one.
You can also open an issue in the repo if something does not work as planned.
{tic} will keep improving the existing templates and its internal functionality.

> We have plans to add an update-mechanism in the future - stay tuned for more!

Reviewer: maybe a footnote on what staying tuned means in practice? Watching the repo?

Reviewer: I'd even add a link to GitHub docs about watching.

Do you use releases at all? If so users could release-watch which is less intense.

Author: > Do you use releases at all?

We do but only for "minor" releases, not for "dev" releases.

> I'd even add a link to GitHub docs about watching.

Added some more links: 
https://github.com/ropensci/roweb2/pull/621/commits/3a71b38823994612103e72b84d9f58ff854aa12a

---
slug: "tic-ghactions"
title: Supercharge your GitHub Actions Experience with tic
# Delete the package_version line below if your post is not about a package
package_version: 0.5.0
authors:
  - Patrick Schratz
# Set the date below to the publication date of your post
date: 2020-02-26
# Set categories to technotes if this is a tech note
categories: technotes
# Leave topicid blank below; will be set by editor
topicid:
# Minimal tags for a post about a community-contributed package 
# that has passed software peer review are listed below
# Consult the Technical Guidelines for information on choosing other tags
tags:
  - R
  - GitHub
  - GitHub Actions
  - Continuous Integration
# The summary below will be used by e.g. Twitter cards

> description: "tic aims to simplify CI related tasks for R projects (not just packages by the way) and now provides supercharged support for GitHub Actions. By enhancing the good work of [r-lib/actions](https://github.com/r-lib/actions) tic provides a DSL for R for CI tasks."

Author: I've added an explanation for the acronym DSL and omitted the suggested sentence.

---

> "Continuous Integration" (CI) has become a standard for proper software development.

Author: 👍

Checking the integrity of software after changes have been made is essential to ensure its proper functionality.

> Also, CI helps catching problems introduced by dependencies early when executed on a regular basis (usually done via scheduled [CRON](https://tecadmin.net/crontab-in-linux-with-20-examples-of-cron-schedule/) runs).

Reviewer 2: ```suggestion
Also, CI helps catch problems introduced by dependencies early when executed on a regular basis (usually done via scheduled [CRON](https://tecadmin.net/crontab-in-linux-with-20-examples-of-cron-schedule/) runs).
```

> Multiple professional providers exist ([Travis CI](https://travis-ci.com/), [Appveyor CI](https://www.appveyor.com/), [Circle CI](https://circleci.com/), etc.) which offer CI services to the public.

Reviewer 2: ```suggestion
Multiple professional providers exist ([Travis CI](https://travis-ci.com/), [AppVeyor CI](https://www.appveyor.com/), [Circle CI](https://circleci.com/), etc.) which offer CI services to the public.
```

Most services offer limited free build contingents for open-source projects.

> However, it has not been easy for the R community to jump on the train right away as most providers did not support the language natively on their side.

Reviewer 2: ```suggestion
However, it has not been easy for the R community to jump on the train right away as most providers did not support the R language natively on their side.
```

Therefore, the R community started to create their own configurations to quickly automate the checking of R packages on [Travis CI](https://travis-ci.com/).
In fact, the first commit of the official `r.rb` build script was made on [Feb 1st 2015 by Craig Citro](https://github.com/travis-ci/travis-build/commit/c697bb2240cfc1abb92a95f57d2e72c151104431).

> Since then 21 unique contributors maintained this configuration under the meta-lead of [Jim Hester](https://github.com/jimhester) and [Jeroen Ooms](https://github.com/jeroen) and many thousand R package maintainers jumped onto the train.

Reviewer 2: ```suggestion
Since then 21 unique contributors have maintained this configuration under the meta-lead of [Jim Hester](https://github.com/jimhester) and [Jeroen Ooms](https://github.com/jeroen) which allowed many thousands of R package maintainers to jump onto the train.
```

However, [Travis CI](https://travis-ci.com/) has it limits and therefore people started to try out other providers such as [Appveyor CI](https://www.appveyor.com/) or [Circle CI](https://circleci.com/).
Besides having to start from scratch, all providers come with their own YAML keys and ways to set up deployment permissions.

With the aim to simplify these processes for users, the tic package was created.
It is aimed to provide a CI-agnostic way to check R packages.
But not just R packages! 
It also aims to simplify the creation of blogs, websites, [drat](https://github.com/eddelbuettel/drat) repositories and other things related to CI activities.

By providing CI YAML templates which forward their actions to a centralized R script (`tic.R`), the same setup will trigger identical actions on different CI systems.
The user does not need to worry about setting up the YAML file for a specific provider but just get familiarized with the "macro" approach of tic.

> [Macros](https://docs.ropensci.org/tic/articles/tic.html#macros) are sequences of steps for popular actions often requested by people in the R community for R related projects.

Reviewer 2: ```suggestion
[tic macros](https://docs.ropensci.org/tic/articles/tic.html#macros) are sequences of steps for popular actions often requested by people in the R community for R related projects.
```

> Furthermore tic aims to simplify the tedious process of setting up deployment permissions for a repo.

Reviewer 2: ```suggestion
Furthermore, tic aims to simplify the tedious process of setting up deployment permissions for a repo.
```

("Deployment" means that a CI run is allowed to perform a `git` push to the repository.)
This is achieved by its helper packages such as circle, ghactions or travis which perform API calls to the respective CI providers.

### GitHub Actions - the new kid on the block

> GitHub Actions (beta) is around since about one year.

Reviewer 2: ```suggestion
GitHub Actions (beta) has been around for about one year.
```

The official release happened in [November 2019](https://siliconangle.com/2019/11/13/github-universe-announcements-bring-bevy-updates-developers/).

> The hype after the launch was huge: People are used to expect properly developed products from GitHub and having an integrated CI solution without the overhead of connecting to a third party provider sounded promising.

Reviewer 2: ```suggestion
The hype after the launch was huge: People expect properly developed products from GitHub and having an integrated CI solution without the overhead of connecting to a third party provider sounded promising.
```

> Indeed, GitHub Actions fulfilled this promise: It is way easier to get started with CI now: 

Reviewer 2: Avoid two colons in a row
```suggestion
Indeed, GitHub Actions fulfilled this promise. It is way easier to get started with CI now: 
```

- No extra account needs to be created for a third-pary service
- No restrictions on the available build platforms
- No issues related to webhooks of third-party providers
- CI settings configurable via the repository settings

However, R was again not among the languages with native support/examples, so it was not easy to get started.
After a first shot from [Max Held](https://github.com/maxheld83) via the [ghactions](https://github.com/maxheld83/ghactions) package, [Jim Hester](https://github.com/jimhester) sat down again and created the [setup-r](https://github.com/r-lib/actions/tree/master/setup-r) actions which installs R on all avaialable platforms with the option to specify different R versions.

> While this action is already usable, its still in its early days and will mature over time.

Reviewer 2: ```suggestion
While the action package is already usable, it's still in its early days and will mature over time.
```

Author: It's not really an R package (even though it has a DESC).
I would like to refer to it as a GitHub action rather than an R package (rephrased manually).

In tic we use [r-lib/setup-r](https://github.com/r-lib/actions/tree/master/setup-r) as the base combined with [r-lib/setup-pandoc](https://github.com/r-lib/actions/tree/master/setup-pandoc) and [r-lib/setup-tinytex](https://github.com/r-lib/actions/tree/master/setup-tinytex).
We apply some opinionated changes on top of which we think enhance the CI setup substantially.

> In the following we explain these changes to be as transparent as possible.

Reviewer 2: ```suggestion
In the following sections explain these changes to be as transparent as possible.
```

#### Caching

> By default [r-lib/actions](https://github.com/r-lib/actions) caches all direct dependencies of the package which is being tested on a successful run.

Author: 👍

This is done via action [actions/cache](https://github.com/actions/cache) by hashing a list of all package dependencies:

```{yml}
- name: Query dependencies
  run: |
    install.packages('remotes')
    saveRDS(remotes::dev_package_deps(dependencies = TRUE), "depends.Rds", version = 2)
  shell: Rscript {0}

- name: Cache R packages
  if: runner.os != 'Windows'
  uses: actions/cache@v1
  with:
    path: ${{ env.R_LIBS_USER }}
    key: ${{ runner.os }}-r-${{ matrix.config.r }}-${{ hashFiles('depends.Rds') }}
    restore-keys: ${{ runner.os }}-r-${{ matrix.config.r }}-
```

> tic instead caches the complete R library of the run **on every run** by using a forked version of [actions/cache](https://github.com/actions/cache): [pat-s/always-upload-cache](https://github.com/pat-s/always-upload-cache).

Author: 👍

Because caches are immutable, tic (needs to) rebuilds the cache **daily**.
If this would not happen, every runner would otherwise always use the first cache forever.
When packages are updated on CRAN, tic would download the cache first but then update those before proceeding to the "script" stage.
After some time all packages from the cache would be outdated, making the cache useless.
Hence, tic rebuilds the cache daily.
Users can change this behavior to their liking by changing the bash code that is used to generate the timestamps used for caching:

```{yml}
- name: "[Cache] Prepare daily timestamp for cache"
  if: runner.os != 'Windows'
  id: date
  run: echo "::set-output name=date::$(date '+%d-%m')"

- name: "[Cache] Prepare weekly timestamp for cache"
  if: runner.os != 'Windows'
  id: datew
  run: echo "::set-output name=datew::$(date '+%Y-%V')"
```

The cache rebuilding happens in a scheduled CRON run daily at 3 am in the morning.
Because tic always uploads the cache even if the build failed ([r-lib/actions](https://github.com/r-lib/actions) only uploads on a successful build), there will always be a cache available for builds which start during the day.

> So even if the build of a repository takes several minutes to install all dependencies, the first commit being made every day will always make use of an existing cache.

Author: I have to temporarily for now because "scheduled" events are not supported atm due to an internal change from GitHub. Will be available in the future though. I'll try to make that clear.

> So even if the build of a repository takes several minutes to install all dependencies, the first commit being made every day will always make use of an existing cache.

Reviewer 2: See my comment above

#### macOS compiler toolchain

macOS is a complicated platform for R.
CRAN does use a custom compiler for creating the binaries for this platform.
One reason is that the default `/usr/bin/clang` does not support `openMP`.
For the current R 3.6.x release a custom version of `clang7` is used while for R 4.x `clang8` will be used.
tic reflects exactly this setting on the macOS-release and macOS-devel runners, eliminating lots of issues when compiling packages from source and giving you the security to have the CRAN setup mirrored in your builds.

#### Addtional tweaks

- tic always runs `R CMD javareconf` to ensure a working installation of rJava.
- tic uses four cores for installing dependencies instead of just running sequentially.
- tic sets the env var `RGL_USE_NULL` to make it possible to install the rgl package if needed.
- tic installs all packages from source on Linux instead of using the R package binaries provided by RStudio.
  This is to mimic the normal behavior on Linux and ensure a proper linking against system libraries.
    
### Getting Started

> If you are triggered now to try out tic and GitHub Actions, here is how to get started (alternatively you can also run `tic::use_tic()` to start an interactive wizard guiding you through the setup process):

Reviewer 2: ```suggestion
If you are triggered now to try out tic and GitHub Actions, here is how to get started (alternatively you can also run `tic::use_tic()` to start an interactive wizard which will guide you through the setup process):
```

```{r, eval=FALSE}
library("tic")

use_ghactions_deploy() # (optional) setup deployment from builds
use_ghactions_yml() # use deploy = TRUE to deploy from builds
tic::use_tic_r(repo_type = "package", deploy_on = "ghactions") # add tic.R
use_tic_badge("ghactions") # (optional) add a badge to README.md/README.Rmd
```

> After that, commit `.github/workflows/main.yml` and `tic.R` and have a look at the "Actions" pane in the repository.

Reviewer 2: ```suggestion
After that, commit `.github/workflows/main.yml` and `tic.R` and have a look at the "Actions" pane in the GitHub repository.
```

For a full example of the resulting YAML file you can have a look at our example repository for packages [tic.package](https://github.com/ropenscilabs/tic.package/blob/25fe68d3d874a671e8a1350dfd47789d7dd19105/.github/workflows/main.yml).

### Summary

tic aims to simplify CI related tasks for R projects (not just packages by the way) and now provides supercharged support for GitHub Actions.
By enhancing the good work of [r-lib/actions](https://github.com/r-lib/actions) tic provides a DSL for R for CI tasks.

> If you encounter problems, please have a look at the vignettes of tic, especially the ["FAQ"](https://docs.ropensci.org/tic/articles/faq.html) one.

Reviewer 2: ```suggestion
If you encounter problems, please have a look at the vignettes of tic, especially the ["FAQ"](https://docs.ropensci.org/tic/articles/faq.html).
```

You can also open an issue in the repo if something does not work as planned.
tic will keep improving the existing templates and its internal functionality[^1].

We have plans to add an update mechanism in the future to simplify staying up-to-date with the latest template changes upstream - stay tuned!

> [^1]: Since tic is not on CRAN yet, you have to watch the development on GitHub for now to stay up-to-date.

Reviewer 2: ```suggestion
[^1]: Since tic is not on CRAN yet, you have to watch the development on [GitHub](https://github.com/ropensci/tic) for now to stay up-to-date.
```

Author: merged manually

> [^1]: Since tic is not on CRAN yet, you have to watch the development on GitHub for now to stay up-to-date.

Reviewer 2: Consider adding an Acknowledgements section with links to the review repository for tic and reviewers author pages.

Author: I haven't done so because it's a technote and not a package announcement - should I still do it?

Reviewer 2: Yes please :grin: It's always nice to showcase the review process and the reviewers, especially since this is the first post about tic on rOpenSci. In addition, I always find my confidence in a package goes up if I know that it has been officially reviewed and vetted by other people not part of the project.

---
slug: "tic-ghactions"
title: Supercharge your GitHub Actions Experience with tic
# Delete the package_version line below if your post is not about a package
package_version: 0.5.0
authors:
  - Patrick Schratz
# Set the date below to the publication date of your post
date: 2020-02-26
# Set categories to technotes if this is a tech note
categories: technotes
# Leave topicid blank below; will be set by editor
topicid:
# Minimal tags for a post about a community-contributed package 
# that has passed software peer review are listed below
# Consult the Technical Guidelines for information on choosing other tags
tags:
  - R
  - GitHub
  - GitHub Actions
  - Continuous Integration
  - software-peer-review
  - tic
  - community
# The summary below will be used by e.g. Twitter cards

> description: "tic aims to simplify Continuous Integration (CI) related tasks for R projects (not just packages by the way) and now provides supercharged support for GitHub Actions. By enhancing the good work of [r-lib/actions](https://github.com/r-lib/actions) tic provides a DSL for R for CI tasks."

Reviewer 2: After checking the Twitter Card validator, I think we need to shorten this a bit, to make sure the important part (support for GitHub actions) is visible.
```suggestion
description: "tic simplifies CI tasks for R projects and now provides supercharged support for GitHub Actions. By enhancing the good work of [r-lib/actions](https://github.com/r-lib/actions) tic provides a DSL for R for CI tasks."
```

[tic macros](https://docs.ropensci.org/tic/articles/tic.html#macros) are sequences of steps for popular actions often requested by people in the R community for R related projects.

Furthermore, tic aims to simplify the tedious process of setting up deployment permissions for a repo.
("Deployment" means that a CI run is allowed to perform a `git` push to the repository.)
This is achieved by its helper packages such as circle, ghactions or travis which perform API calls to the respective CI providers.

### GitHub Actions - the new kid on the block

GitHub Actions (beta) has been around for about one year.
The official release happened in [November 2019](https://siliconangle.com/2019/11/13/github-universe-announcements-bring-bevy-updates-developers/).
The hype after the launch was huge: People expect properly developed products from GitHub and having an integrated CI solution without the overhead of connecting to a third party provider sounded promising.

Indeed, GitHub Actions fulfilled this promise. It is way easier to get started with CI now: 

> - No extra account needs to be created for a third-pary service

Reviewer 2: ```suggestion
- No extra account needs to be created for a third-party service
```

So even if the build of a repository takes several minutes to install all dependencies, the first commit being made every day will always make use of an existing cache.

#### Installation of dependencies

tic uses `ccache` on macOS and Linux to speed up R package installations.
This tool helps speeding up the installation of compiled packages by loading the C code parts from the cache.
tic rebuilds the `ccache` only **once a week** making it possible that the CRON builds in the morning can make use of the `ccache` builds when rebuilding the R package cache.

#### macOS compiler toolchain

macOS is a complicated platform for R.
CRAN does use a custom compiler for creating the binaries for this platform.
One reason is that the default `/usr/bin/clang` does not support `openMP`.
For the current R 3.6.x release a custom version of `clang7` is used while for R 4.x `clang8` will be used.
tic reflects exactly this setting on the macOS-release and macOS-devel runners, eliminating lots of issues when compiling packages from source and giving you the security to have the CRAN setup mirrored in your builds.

> #### Addtional tweaks

Reviewer 2: ```suggestion
#### Additional tweaks
```

- tic always runs `R CMD javareconf` to ensure a working installation of rJava.
- tic uses four cores for installing dependencies instead of just running sequentially.
- tic sets the env var `RGL_USE_NULL` to make it possible to install the rgl package if needed.
- tic installs all packages from source on Linux instead of using the R package binaries provided by RStudio.
  This is to mimic the normal behavior on Linux and ensure a proper linking against system libraries.
    

> ### Getting Started

Reviewer 2: ```suggestion
### Getting started
```

  - GitHub
  - GitHub Actions
  - Continuous Integration
  - software-peer-review
  - tic
  - community
# The summary below will be used by e.g. Twitter cards
description: "tic aims to simplify Continuous Integration (CI) related tasks for R projects (not just packages by the way) and now provides supercharged support for GitHub Actions. By enhancing the good work of [r-lib/actions](https://github.com/r-lib/actions) tic provides a DSL for R for CI tasks."
---

> Continuous Integration (CI) has become a standard for proper software development.

Reviewer 2: Could you define Continuous Integration here? Not just what the acroynym means, but what the term means as well. That'll help beginner R programmers who may be less familiar with this field.

Author: I've added a link to wikipedia. Is that sufficient? I am personally always skeptical whether such terms should be explained again from the ground up in a blog post (I know I might face different opinions here).

Reviewer 2: I'm happy with this compromise :slightly_smiling_face:

("Deployment" means that a CI run is allowed to perform a `git` push to the repository.)

> This is achieved by its helper packages such as {circle}, {ghactions} or {travis} which perform API calls to the respective CI providers.

Reviewer: in our style guide, package names need to be written without curly braces or any special formatting. :-)

Author: Thanks, I'll consult it again before submitting for review.

## GitHub Actions - the new kid on the block

> GitHub Actions is around since about one year.

Reviewer: Not really in its current version? See e.g. https://siliconangle.com/2019/11/13/github-universe-announcements-bring-bevy-updates-developers/

Author: Ok then let's refer to the official release date :)

The hype after the launch was huge: People are used to expect properly developed products from GitHub and having an integrated CI solution without the overhead of connecting to a third party provider sounded promising.

Indeed, GitHub Actions fulfilled this promise: It is way easier to get started with CI now: 

- No extra account needs to be created
- No restrictions on the available build platforms
- No issues related to webhooks for third-party providers
- CI settings configurable via the repository settings

However, R was again not among the languages with native support/examples, so it was not easy to get started.
After a first shot from [Max Held]() via the [{ghactions}]() package, [Jim Hester]() sat down again and created an "action" which installs R on all platforms with the option to specify different R versions.

While this action is already usable, its still quite young and will mature over time.
In {tic} we use [r-lib/actions]() as the base and apply some opinionated changes of which we think enhance the CI runs substantially.
In the following we explain these changes to be as transparent as possible.

### Caching

> By default [r-lib/actions]() caches all direct dependencies of the package which is being tested on a successful run.

Reviewer: How does this happen? I thought one needed an actual caching step?

Author: We have an official caching step - I'll link the one from our template.

{tic} caches the complete R library of the run.
This has the advantage that also packages which are needed for side-effects (like building a pkgdown site or running {codecov}) are cached.
Because caches are immutable, {tic} rebuilds the cache **daily**.
If not, every runner would otherwise always use the same cache forever.
If packages are updated on CRAN, {tic} would update those before proceeding to the "script" stage.

> After some time all packages would be outdated and the cache quite useless.

Reviewer: Can one change this to say weekly, monthly etc?

Author: Yes, I'll add a sentence how this can be done.

Reviewer: You mean the sentence starting with "Users can change this behavior to their liking"? If so can you add "Users can change this behavior to their liking (frequency, otherthing)" or so?

Author: Yeah, tried to make it more clear again: https://github.com/ropensci/roweb2/pull/621/commits/980d20d152442e980c9f79e7b12e639c08ee9832

However, R was again not among the languages with native support/examples, so it was not easy to get started.
After a first shot from [Max Held](https://github.com/maxheld83) via the [ghactions](https://github.com/maxheld83/ghactions) package, [Jim Hester](https://github.com/jimhester) sat down again and created the [setup-r](https://github.com/r-lib/actions/tree/master/setup-r) actions which installs R on all avaialable platforms with the option to specify different R versions.

> While this GitHub action is already usable, it's still in its early days and will mature over time.

Author: No, this is correct 👍

```

The cache rebuilding happens in a scheduled CRON run daily at 3 am in the morning.

> Because tic always uploads the cache even if the build failed ([r-lib/actions](https://github.com/r-lib/actions) only uploads on a successful build), there will always be a cache available for builds which start during the day[^1].

Reviewer 2: Hmm, tricky... Perhaps be a bit more explicit in the text what the problem is, and clarify that it's not tic's problem :) Just in case readers jump right in and are then confused that something doesn't work as expected.
Once again, this is assuming I've understood the context!
 
```suggestion
Because tic always uploads the cache even if the build failed ([r-lib/actions](https://github.com/r-lib/actions) only uploads on a successful build), there will always be a cache available for builds which start during the day. Unfortunately at the time of writing, GitHub actions do not support CRON events ("scheduled"). However, this is only a temporary state and will most likely change in the future (see [this issue](https://github.com/actions/cache/issues/63) for more information).
```

Author: I've clarified again - let me know if is more clear now.
